File: find-minimum-in-rotated-sorted-array-ii/find-minimum-in-rotated-sorted-array-ii.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
        int low = 0;
        int high = nums.size()-1;
        int mid = 0;
        
        while (low<high) {
            mid = low + (high-low)/2;
            
            if (nums[mid]>nums[high])
                // mid is surely not the answer. Hence mid+1
                low = mid+1;
            else if (nums[mid]<nums[high])
                // mis may be the answer still. Hence not mid-1.
                high = mid;
            else
                // assuming nums[mid]==nums[high]. Answer can be anywhere in between or just either of them. Hence move 1 by 1 index. 
                high-=1;
        }
        
        // the low pointer points to the minimum element in array
        return nums[low];
    }
};

File: find-minimum-in-rotated-sorted-array-ii/find-minimum-in-rotated-sorted-array-ii.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        low, mid, high = 0, 0, len(nums)-1
        
        while low<high:
            mid = low + (high-low)//2
            
            if nums[mid] > nums[high]:
                # mid is not the answer
                # smallest number lies between (mid+1) and (high)
                low = mid+1
            elif nums[mid] < nums[high]:
                # mid might be a potential answer
                # but we can now narrow down the search to (low) and (mid)
                high = mid
            else:
                # mid to high is basically useless, as they are equal
                # so we now narrow down 1 step at a time. 
                high -= 1
                
        return nums[low]